kvstore/bolt/bolt.go

```diff
@@ -0,0 +1,280 @@
+	return prefix
+}
+
+func copyBytes(source []byte) []byte {
+	cpy := make([]byte, len(source))
+	copy(cpy, source)
+	return cpy
+}
+
+func (s boltStore) iterate(prefixes []kvstore.KeyPrefix, copyValues bool, kvConsumerFunc kvstore.IteratorKeyValueConsumerFunc) error {
+	return s.instance.View(func(tx *bbolt.Tx) error {
+		b := tx.Bucket(s.bucket)
+		if b == nil {
+			return nil
+		}
+		c := b.Cursor()
+
+		if len(prefixes) > 0 {
+			prefix := buildPrefixedKey(prefixes)
+			for k, v := c.Seek(prefix); k != nil && bytes.HasPrefix(k, prefix); k, v = c.Next() {
+				value := v
+				if copyValues {
+					value = copyBytes(v)
+				}
+				if !kvConsumerFunc(copyBytes(k), value) {
+					break
+				}
+			}
+			return nil
+		}
+
+		for k, v := c.First(); k != nil; k, v = c.Next() {
+			value := v
+			if copyValues {
+				value = copyBytes(v)
+			}
+			if !kvConsumerFunc(copyBytes(k), value) {
+				break
+			}
+		}
+		return nil
+	})
+}
+
+func (s *boltStore) Iterate(prefixes []kvstore.KeyPrefix, kvConsumerFunc kvstore.IteratorKeyValueConsumerFunc) error {
+	return s.iterate(prefixes, true, kvConsumerFunc)
+}
+
+func (s *boltStore) IterateKeys(prefixes []kvstore.KeyPrefix, consumerFunc kvstore.IteratorKeyConsumerFunc) error {
+	// same as with values but we simply don't copy them
+	return s.iterate(prefixes, false, func(key kvstore.Key, _ kvstore.Value) bool {
+		return consumerFunc(key)
+	})
+}
+
+func (s *boltStore) Clear() error {
+	return s.instance.Update(func(tx *bbolt.Tx) error {
+		if tx.Bucket(s.bucket) == nil {
+			return nil
+		}
+		return tx.DeleteBucket(s.bucket)
+	})
+}
+
+func (s *boltStore) Get(key kvstore.Key) (kvstore.Value, error) {
+	var value []byte
+	if err := s.instance.View(func(tx *bbolt.Tx) error {
+		b := tx.Bucket(s.bucket)
+		if b == nil {
+			return nil
+		}
+		v := b.Get(key)
+		if v != nil {
+			value = copyBytes(v)
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	if value == nil {
+		return nil, kvstore.ErrKeyNotFound
+	}
+	return value, nil
+}
+
+func (s *boltStore) Set(key kvstore.Key, value kvstore.Value) error {
+	return s.instance.Update(func(tx *bbolt.Tx) error {
+		b, err := tx.CreateBucketIfNotExists(s.bucket)
+		if err != nil {
+			return err
+		}
+		return b.Put(key, value)
+	})
+}
+
+func (s *boltStore) Has(key kvstore.Key) (bool, error) {
+	var has bool
+	err := s.instance.View(func(tx *bbolt.Tx) error {
+		b := tx.Bucket(s.bucket)
+		if b == nil {
+			return nil
+		}
+		has = b.Get(key) != nil
+		return nil
+	})
+	return has, err
+}
+
+func (s *boltStore) Delete(key kvstore.Key) error {
+	return s.instance.Update(func(tx *bbolt.Tx) error {
+		b, err := tx.CreateBucketIfNotExists(s.bucket)
+		if err != nil {
+			return err
+		}
+		if err := b.Delete(key); err != nil {
+			return kvstore.ErrKeyNotFound
+		}
+		return nil
+	})
+}
+
+func (s *boltStore) DeletePrefix(prefix kvstore.KeyPrefix) error {
+	return s.instance.Update(func(tx *bbolt.Tx) error {
+		b, err := tx.CreateBucketIfNotExists(s.bucket)
```

Comment:
Do we have to create the bucket at deletion?

Reply:
Yes, you need to operate on the bucket, no way to delete a key without having the bucket